sshd-core/src/main/java/org/apache/sshd/common/io/nio2/Nio2Session.java

```diff
@@ -309,10 +309,16 @@ protected void handleReadCycleCompletion(
                     }
                 }
             } else {
+                // In case remote socket has reached eof, we have to close the session
+                // gracefully (eg. TcpipServerChannel need to send and ordered
+                // SSH_MSG_CHANNEL_EOF and SSH_MSG_CHANNEL_CLOSE sequence).
+                boolean immediately = false;
+
                 if (log.isDebugEnabled()) {
                     log.debug("handleReadCycleCompletion({}) Socket has been disconnected (result={}), closing IoSession now", this, result);
                 }
-                close(true);
+
+                close(immediately);
```

Comment:
This decision seems too early - `Nio2Session` does not know it is invoked as part of the channel or session or whatever. Is there some way we can limit this to just the `TcpipServerChannel` ? Somehow detect it is being invoked in that context ?

Reply:
At the moment it's not clear to me what happens using `MINASession`, which looks like a complete different approach. I understand you comments on releation between `Nio2Session` and `TcpipServerChannel`. 
Anyway, I think that channel EOF should be consided a 'normal' condition so that a 'graceful' shoutdown should be trigered.
As far as I understood, the relation between `Nio2Session` and 'TcpipServerChannel` seems to be 'embedded' inside the `ioHandler`. May be we could change 'IoHandler' interface and transform `sessionClosed(IoSession session)` into `sessionClosed(IoSession session, boolean immediatly)`, so to let specific `ioHandler` do the right job inside `handleReadCycleCompletion`.

Like:

```
            } else {
                if (log.isDebugEnabled()) {
                    log.debug("handleReadCycleCompletion({}) Socket has been disconnected (result={}), closing IoSession now", this, result);
                }
                IoHandler handler = getIoHandler();
                handler.sessionClosed(this, false);

                close(true);
            }
```